From 749234008a085c116d7170717c027aba1686e1f7 Mon Sep 17 00:00:00 2001
From: Andreas Schultz <andreas.schultz@travelping.com>
Date: Wed, 28 Apr 2021 16:54:50 +0200
Subject: [PATCH 3/3] diameter: fix delays in gen_sctp test suite

Also add warning note about delay with enable Nagle alogrithm on
SCTP.
---
 lib/diameter/doc/src/diameter_sctp.xml        | 5 +++++
 lib/diameter/test/diameter_gen_sctp_SUITE.erl | 3 ++-
 2 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/lib/diameter/doc/src/diameter_sctp.xml b/lib/diameter/doc/src/diameter_sctp.xml
index bbb9fbab9d..39a9efd329 100644
--- a/lib/diameter/doc/src/diameter_sctp.xml
+++ b/lib/diameter/doc/src/diameter_sctp.xml
@@ -170,6 +170,11 @@ the buffer size.</p>
 An small send buffer may result in outgoing messages
 being discarded: set the &inet; option <c>sndbuf</c> to increase
 the buffer size.</p>
+
+<p>
+Having the Nagle algorithm enabled (the default) can cause unexpected delays
+for message oriented protocols like DIAMETER.
+Set the &inet; option <c>sctp_nodelay</c> to <c>false</c> to disable it.</p>
 </warning>
 
 </desc>
diff --git a/lib/diameter/test/diameter_gen_sctp_SUITE.erl b/lib/diameter/test/diameter_gen_sctp_SUITE.erl
index d546e77cde..9d9840a855 100644
--- a/lib/diameter/test/diameter_gen_sctp_SUITE.erl
+++ b/lib/diameter/test/diameter_gen_sctp_SUITE.erl
@@ -448,7 +448,8 @@ open(Opts) ->
     gen_sctp:open([{ip, ?ADDR}, {port, 0}, {active, false}, binary,
                    {sctp_initmsg, #sctp_initmsg{num_ostreams = ?STREAMS,
                                                 max_instreams = ?STREAMS}},
-                   {recbuf, 1 bsl 16}, {sndbuf, 1 bsl 16}
+                   {recbuf, 1 bsl 16}, {sndbuf, 1 bsl 16},
+                   {sctp_nodelay, true}
                    | Opts]).
 
 %% report/2
-- 
2.30.2

